### DISCUSSION TITLE
Zustand with Preact

### DISCUSSION THREAD

**Discussion by ndrean** [pmndrs/zustand#1697] [General]
Probably an ongoing question but I couldn't find an answer, so sorry in advance!

If I install Zustand in a Preact app, it brings React as seen with `npm ls react` so my build suddenly inflates. Could there be Preact-based Zustand ?

**Comment by dai-shi** [comment]
I've never tried personally. At least, react is marked as optional. 

https://github.com/pmndrs/zustand/blob/962f83e01599257aa6eceaaa993965a0e296d359/package.json#L227-L229

**Reply by ndrean** [reply]
Zustand brings in React & ReactDOM (they appear in my node_modules). I tried to add the package <https://www.npmjs.com/package/@preact/compat> , but this does not help either. React is still brought into the node_modules.

```json
"dependencies": {
    "history": "^5.3.0",
    "preact": "^10.11.3",
    "react": "npm:@preact/compat@^17.1.2",
    "react-dom": "npm:@preact/compat@^17.1.2",
    "universal-router": "^9.1.0",
    "zustand": "^4.3.6"
  },
```

**Reply by dai-shi** [reply]
that's weird. zustand doesn't depend on react-dom anyway.

**Reply by ndrean** [reply]
I forgot to erase ReactDOM, sorry for the confusion. Only React. Anyway, I was hoping to make a much slimmer build with Preact and Zustand. If it's not possible, then at least Preact saves a bit without bringing in ReactDOM.

**Reply by dai-shi** [reply]
Erase react too. It's optional. As long as preact/compat provides a compat layer, it should work.

**Comment by Simon-Tang** [comment]
Adding this into my `package.json` seems to work:

```
  "overrides": {
    "zustand": {
      "react": "npm:preact"
    }
  }
```

After reinstalling Zustand, `npm ls react` confirms that `react` has been overridden. `node_modules/react/` contains Preact code instead.

```
npm ERR! code ELSPROBLEMS
npm ERR! invalid: react@npm:preact@10.16.0 /Users/.../node_modules/react
lofi-fx@0.0.0 /Users/...
└─┬ zustand@4.3.9 overridden
  ├── react@npm:preact@10.16.0 overridden
  └─┬ use-sync-external-store@1.2.0
    └── react@npm:preact@10.16.0 deduped invalid: "^16.8.0 || ^17.0.0 || ^18.0.0" from node_modules/use-sync-external-store
```

However, with `npm 9.5.0` the `overrides` feature still seems iffy. This only worked for me with the `npm:` prefix which [isn't documented](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides), and I wasn't able to specify a specific version of `preact`. If I try `preact^10.16.0` or `npm:preact^10.16.0` instead of `npm:preact`:

```
npm ERR! code EINVALIDTAGNAME
npm ERR! Invalid tag name "preact^10.16.0" of package "preact^10.16.0": Tags may not have any characters that encodeURIComponent encodes.
```

So this might be an issue with npm that will need to be addressed, or you'd end up with Zustand always using the latest Preact version.

**Reply by dbritto-dev** [reply]
That's due to when you use `mpm:` what you are doing is installing package from `GitHub` and you need to use `npm:preact@10` rather than `npm:preact^10`

**Reply by Simon-Tang** [reply]
Thanks @dbritto-dev, you're right about the `npm:preact@10`, that syntax does work. I'm not sure why I'm getting packages from GitHub, but you're also right again, that is what's happening in my case.

I hope the `overrides` workaround helps, to at least reduce the bundle size, but I second the question about whether it would be possible to have vanilla Zustand without React.

**Reply by dbritto-dev** [reply]
@Simon-Tang you only need to import it like this `zustand/vanilla` and that's it
